lib/govuk_publishing_components/app_helpers/asset_helper.rb

```diff
@@ -0,0 +1,155 @@
+      # added and removed from every single rendering application.
+      COMPONENT_CSS_PATHS = %w[
+        accordion
+        action-link
+        attachment
+        attachment-link
+        back-link
+        big-number
+        breadcrumbs
+        button
+        cards
+        character-count
+        checkboxes
+        contents-list
+        contextual-guidance
+        contextual-sidebar
+        cookie-banner
+        copy-to-clipboard
+        date-input
+        details
+        devolved-nations
+        document-list
+        emergency-banner
+        error-alert
+        error-message
+        error-summary
+        feedback
+        fieldset
+        file-upload
+        govspeak
+        govspeak-html-publication
+        heading
+        hint
+        image-card
+        input
+        inset-text
+        intervention
+        inverse-header
+        label
+        layout-footer
+        layout-for-admin
+        layout-for-public
+        layout-header
+        layout-super-navigation-header
+        lead-paragraph
+        metadata
+        modal-dialogue
+        notice
+        organisation-logo
+        panel
+        phase-banner
+        previous-and-next-navigation
+        print-link
+        radio
+        related-navigation
+        reorderable-list
+        search
+        select
+        share-links
+        show-password
+        signup-link
+        single-page-notification-button
+        skip-link
+        step-by-step-nav
+        step-by-step-nav-header
+        step-by-step-nav-related
+        subscription-links
+        success-alert
+        summary-list
+        table
+        tabs
+        textarea
+        title
+        translation-nav
+        warning-text
+      ].map { |component|
+        "govuk_publishing_components/components/_#{component}.css"
+      }.freeze
+
+      def initialize
+        # Set up the lists of gem components, app components, and views.
+        @gem_component_stylesheets_being_used = []
+        @app_component_stylesheets_being_used = []
+        @view_stylesheets_being_used = []
+
+        # This list includes components already included in Static; taken from
+        # https://github.com/alphagov/static/blob/198a598682df40ce4a2c3c286c06244297c18cf0/app/assets/stylesheets/application.scss
+        #
+        # This is used to dedupe stylesheets without needing the full path.
+        @stylesheet_list = %w[
+          breadcrumbs
+          button
+          error-message
+          heading
+          hint
+          input
+          label
+          search
+          skip-link
+          textarea
+          title
+
+          cookie-banner
+          feedback
+          layout-footer
+          layout-for-public
+          layout-header
+          layout-super-navigation-header
+        ]
+      end
+
+      # Used to add a component that exists in the gem to the list
+      def add_gem_component(component)
+        unless is_already_used?(component)
```

Comment:
This check and subsequent ones looks like it has some problems:

- If there is any cases of a stylesheet having the same name it wouldn't include it. E.g if you had a view named and component sharing the same name the last added would be rejected
- if you had a view that had the same name as any of the reserved ones for static it'd not be included
- if you had an app that didn't use static you'd never be able to use the static components (I imagine it'd be better to configure this with an initialiser to state whether an app uses static)

Reply:
> Am I correct in understanding that the add_gem_component, add_component and add_view methods will be used by the rendering applications?

Yes, although I don't think they need to be called in a specific order, will test this further to confirm if this is correct.

> Does that mean that it is expected that the stylesheet for the gem will be the one used even if a view in a rendering application has one with the same name?

I think this would depend on the order in which the component stylesheet paths are added to `@stylesheet_list`, with the current implementation, the below scenario is possible:

- `views/_COMPONENT_NAME.css` is used and `_COMPONENT_NAME` added to`@stylesheet_list` first
- The view template renders a component with the same name, either from the gem or the app itself, `components/_COMPONENT_NAME.css`
- The `is_already_used?` method will stop this component name from being added to `@stylesheet_list`, as it is already in included
```
def is_already_used?(component)
        @stylesheet_list.include?(component)
end
```
- The stylesheet for `components/_COMPONENT_NAME.css` would not be included on the frontend app and the component would not render correctly

Will try and add a test for this scenario.

> Are there any cases where there are stylesheets with the same name?

I haven't come across any examples of this, but will also check this further.

Reply:
One possible fix for this is to update the methods so that the full path is added to `@stylesheet_list`.

**From**

      ```Ruby
def add_view(view)
        unless is_already_used?(view)
          @view_stylesheets_being_used << "views/_#{view}.css"
          @stylesheet_list << view
        end
      end
```

**To**

```Ruby
def add_view(view)
        unless is_already_used?(view)
          @view_stylesheets_being_used << "views/_#{view}.css"
          @stylesheet_list << "views/_#{view}.css"
        end
      end
```

Reply:
> - If there is any cases of a stylesheet having the same name it wouldn't include it. E.g if you had a view named and component sharing the same name the last added would be rejected.
> - if you had a view that had the same name as any of the reserved ones for static it'd not be included

This potential issue was fixed by including the path to the stylesheet as well as the stylesheet component name.